base/main_test_bucket_pool.go

```diff
@@ -0,0 +1,650 @@
+package base
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/couchbase/gocb"
+	"github.com/couchbaselabs/walrus"
+	"github.com/pkg/errors"
+)
+
+// TestBucketPool is a global instance of a GocbTestBucketPool used to manage a pool of buckets for integration testing.
+var TestBucketPool *GocbTestBucketPool
+
+const (
+	tbpClusterUsername = DefaultCouchbaseAdministrator
+	tbpClusterPassword = DefaultCouchbasePassword
+
+	tbpBucketNamePrefix = "sg_int_"
+
+	// Creates this many buckets in the backing store to be pooled for testing.
+	defaultBucketPoolSize = 10
+	tbpEnvPoolSize        = "SG_TEST_BUCKET_POOL_SIZE"
+
+	defaultBucketQuotaMB = 150
+	tbpEnvBucketQuotaMB  = "SG_TEST_BUCKET_QUOTA_MB"
+
+	// Prevents reuse and cleanup of buckets used in failed tests for later inspection.
+	// When all pooled buckets are in a preserved state, any remaining tests are skipped instead of blocking waiting for a bucket.
+	tbpEnvPreserve = "SG_TEST_BUCKET_POOL_PRESERVE"
+
+	// When set to true, all existing test buckets are removed and recreated, instead of running the bucket readier.
+	tbpEnvRecreate = "SG_TEST_BACKING_STORE_RECREATE"
+
+	// Prints detailed debug logs from the test pooling framework.
+	tbpEnvVerbose = "SG_TEST_BUCKET_POOL_DEBUG"
+)
+
+// GocbTestBucketPool is used to manage a pool of gocb buckets on a Couchbase Server for testing purposes.
+// The zero-value/uninitialized version of this struct is safe to use as Walrus buckets are returned.
+type GocbTestBucketPool struct {
```

Comment:
What do you think about renaming this to remove the Gocb - it's a bit counter-intuitive since it's got built-in walrus support.

Reply:
Makes sense, the external API isn't bound to a specific concrete bucket type.